fix: Ensuring spec file presence prior to webpack-dev-server compilation

tbiethman opened the pull request:
<!-- Thanks for contributing! PLEASE...
- Read our contributing guidelines: https://github.com/cypress-io/cypress/blob/develop/CONTRIBUTING.md 
- Read our Code Review Checklist on coding standards and what needs to be done before a PR can be merged: https://github.com/cypress-io/cypress/blob/develop/CONTRIBUTING.md#Code-Review-Checklist
- Mark this PR as "Draft" if it is not ready for review.
- Make sure you set the correct base branch based on what packages you're changing: https://github.com/cypress-io/cypress/blob/develop/CONTRIBUTING.md#branches
-->

- Closes UNIFY-1697

### User facing changelog
<!-- 
Explain the change(s) for every user to read in our changelog. Examples: https://on.cypress.io/changelog
If the change is not user-facing, write "n/a".
-->

### Additional details
<!-- Examples:
- Why was this change necessary?
- What is affected by this change?
- Any implementation details to explain?
-->

There is a race condition between our webpack compilation and our emission of the `dev-server:specs:changed` event. When a spec file is removed from the file system, we attempt to compile prior to updating our internal `files` property within the plugin, causing the loader to reference files that no longer exist.

I updated the `CypressCTWebpackPlugin` to filter out files in state that no longer exist on the file system during the beforeCompile phase. This prevents both the initial error from being presented at file removal time and allows the file to be immediately executed when re-created.

### How has the user experience changed?
<!-- Provide before and after examples of the change.
Screenshots or GIFs are preferred. -->


https://user-images.githubusercontent.com/1711637/169138230-2e4a3493-0726-4a1c-9eb1-2074fadc9aef.mov


### PR Tasks
<!-- 
These tasks must be completed before a PR is merged.
If a task does not apply, write [na] instead of checking the box.
DO NOT DELETE the PR checklist.
-->

- [X] Have tests been added/updated?
- [n/a] Has the original issue (or this PR, if no issue exists) been tagged with a release in ZenHub? (user-facing changes only)
- [n/a] Has a PR for user-facing changes been opened in [`cypress-documentation`](https://github.com/cypress-io/cypress-documentation)? <!-- Link to PR here -->
- [n/a] Have API changes been updated in the [`type definitions`](https://github.com/cypress-io/cypress/blob/develop/cli/types/cypress.d.ts)?
- [n/a] Have new configuration options been added to the [`cypress.schema.json`](https://github.com/cypress-io/cypress/blob/develop/cli/schema/cypress.schema.json)?

lmiller1990 commented on npm/webpack-dev-server/src/CypressCTWebpackPlugin.ts:
```diff
@@ -68,6 +68,16 @@ export class CypressCTWebpackPlugin {
+
+    // Ensure we don't try to load files that have been removed from the file system
+    // but have not yet been detected by the onSpecsChange handler
+    this.files = (this.files || []).filter((file) => fs.existsSync(file.absolute))
```
I am not sure we want to use `existsSync` here, should we use the async `pathExists` from `fs-extra` and `tapAsync`?

Eg: https://github.com/cypress-io/cypress/compare/tbiethman/fix/unify-1697-ct-file-removed...lmiller1990/fix/unify-1697-ct-file-removed?expand=1

lmiller1990 replied:
https://webpack.js.org/api/compiler-hooks/#beforecompile can work with `tapAsync`

lmiller1990 replied:
Or - is the point we **must** do this async, so we avoid the race condition?

tbiethman replied:
We just need this to execute prior to the compile phase, which should happen either sync or async. Given that we're not actively manipulating the compile params here, we might as well use async and whatever benefit that might get us.

Your example PR was helpful, thank you. I thought about going with a Promise.all rather than running them in series. Something like:

```js
    const foundFiles = await Promise.all(this.files.map(async (file) => {
      try {
        const exists = await fs.pathExists(file.absolute)

        return exists ? file : null
      } catch (e) {
        return null
      }
    }))

    this.files = foundFiles.filter((file) => file !== null)
```

Thoughts?

lmiller1990 replied:
Seems fine 💯

lmiller1990 reviewed (commented):
Just a quick question around sync vs async!

lmiller1990 commented on npm/webpack-dev-server/cypress/e2e/webpack-dev-server.cy.ts:
```diff
@@ -8,4 +8,39 @@ describe('Config options', () => {
   })
+
+  // https://cypress-io.atlassian.net/browse/UNIFY-1697
+  it('filters missing spec files from loader during pre-compilation', () => {
```
They are quite cool, but I wonder if a regression test would be best added in the actual `npm/webpack-dev-server`?

tbiethman commented:
I updated this to use an async beforeCompile implementation, seems to work great. I also added a mocha integration test to get slightly more granular coverage of this behavior. To help with this, I added a new event (`dev-server:compile:done`) to our existing `done` hook so I could more easily test that a failure is not occurring during the compile without relying on a timeout.

lmiller1990 reviewed (approved):
Tests were improved, looks good. @ZachJW34 had previously QA'd so I'll merge this up.
